### PR TITLE
Fix handling of source encoding under Python 3

### PR DESCRIPTION
**Pull request by MatthieuDartiailh**
Previously the parser created a Str node instead of a Bytes node.

### CODE REVIEW

**Comment by tillbey** [comment]
I still had the b-string issue after pip installing this fix. I noticed the lextab and parsetab files (in core/parsing/pare_tab) are not rewritten. Deleting them solved the issue.
But as they are dynamically generated only after importing enaml the first time pip doesn't know about them. They are also not deleted with pip uninstall.
I think this could potentially give issues in upgrading. Can we force rewriting these files? Or generate them during installation so pip is aware of them?

**Comment by MatthieuDartiailh** [comment]
I am actually aware of that issue but never took time to address it. Before parse/lex_tabs were bundled but with so many supported version it is not feasible anymore. Checking the timestamp of the cache versus the source file before loading the cache would solve the issue. I will try to give it a look next week (I am away for the week-end). Furthermore I suspect something is wrong in the handling of b-string and actually the handling of rb is definitively wrong. Will also try to fix next week (but if you want to try you are welcome)
By the way thanks for the quick review !!!

**Comment by MatthieuDartiailh** [comment]
Actually doing some reading about allowed string formats helped here. It turns out that rb is not correct br is. And furthermore ur was removed in Python 3.3 so no need to support it. However that does not fix the issue of the parse/lex tabs. 
Trying to think a bit outside of the situation I am used to I realized that writting those cache on the fly as I was suggesting will not work for non-admin. So I need a setuptools solution...

**Comment by tillbey** [comment]
I thought so, too. I'm not very familiar with setuptools but what I've been reading so far it might be possible to call EnamlBaseLexer/EnamlBaseParser in a custom command class? Just need to find out which command, so that the corresponding packages are installed already.

**Comment by MatthieuDartiailh** [comment]
Actually digging around to understand #20 I discovered that the handling of source files under python 3 is terrible. I will try to fix everything at once but I want to have some proper tests so it will take me so time to get it right hope it does not bother you too much (if you really need it we can merge #17 before getting this one in).

**Comment by tillbey** [comment]
Well, I didn't intend kicking of all this. But I agree it's good to fix those issues now that they came up. Take your time, and tell me if I can be of any help.

**Comment by MatthieuDartiailh** [comment]
I believe the latest changes should cover all cases properly. If you have time test it and some ideas about writing tests for this you are welcome.

**Comment by tillbey** [comment]
Unicode in string literals works, I guess that's most important. But Python 3 also allows unicode identifiers (PEP3131). Should enaml as superset of Python support that, too? See this example I put together: https://gist.github.com/tillbey/674f8a6bdc80d35a40684426bd17591d
Even in the .py file the smiley doesn't work, though the Chinese character does...

**Comment by MatthieuDartiailh** [comment]
Ideally I would say yes. Thanks for catching this and providing test scripts. I will try to see how hard it would be to support it. If it it looks like too much work I will focus on writing the tests and leave this for a following PR. I fear I will quite busy next week so it is sadly possible this will have to wait a bit.

**Comment by tillbey** [comment]
I agree this is less of a priority. I tried to find the source of this not working but I fear it may be the yacc parser itself. That goes to far for me to see if it can be fixed by us.

**Comment by MatthieuDartiailh** [comment]
Actually it it the regex used for the NAME tokens. To fix it we can grab the regex from the tokenize module (tokenise.Name). However because this regex is quite permissive under Python 3 ('\w+'), we need to check it is defined late enough and also to validate it using str.isidentifier. I need to check there are no other issues however.

**Comment by MatthieuDartiailh** [comment]
So the handling of unicode identifier is fixed. But still no automatic tests... @tillbey if you can test manually that would be great (or even write tests).

**Comment by tillbey** [comment]
Sorry, it took some time. I'm neither very familiar with writing tests nor with the string mess. So what I see (under Python 3) that is passing:
1. unicode literals
2. raw unicode literals
3. byte literals
4. raw byte literals (but see below)
5. unicode identifieres

So that's great. But there still is a few cases failing:
1. unicode escapes (\U...) are not converted into unicode characters in u"" or "" literals
2. bytes literals do not convert escape characters (e.g. \n)
3. raw bytes literals don't allow rb"" (SyntaxError) as alternative to br""

**Comment by MatthieuDartiailh** [comment]
Thanks for writing those tests, very helpful. I fixed all issues and included your tests here (and added some more). I will try to add some more tests for python 2 and we will be good go to. It means that file encoding detection based on the magic line won't be tested but I am less concerned about that. Thanks again for your work on this.

**Comment by MatthieuDartiailh** [comment]
I have ported the tests to Python  and actually improved the support for unicode under Python 2. From my end this is good to go. I will let you, @tillbey, make a last review then I will merge and then merge the related PRs #19 and #17.

File: tests/test_encodings.py [hunk 1]
```diff
@@ -66,33 +70,44 @@ def test_raw_bytes_literal1():
...
 
 
 if __name__ == '__main__':
     test_unicode_literal()
     test_unicode_escape_literal()
     test_raw_unicode_literal()
     test_bytes_literal()
-    test_raw_bytes_literal()
+    test_raw_bytes_literal1()
     test_unicode_identifier()
```
Comment (tillbey) [inline tests/test_encodings.py]: This line has to be moved down in the Python 3 extra.

**Review by tillbey — COMMENTED** [review]
Thanks for the honour! I looks great, everything works but one test call. I'm happy I could so easily contribute and that you were so responsive and willing to work these issues out. It really would be a shame to see enaml not being used. So I hope you keep the good work up, maybe until @sccolbert finds some time again.

**Comment by MatthieuDartiailh** [comment]
Thanks for your enthusiasm ! It is nice to work with somebody rather than alone. I addressed your last comment (I am too used to run tests with pytest). So I am merging this, then the two drag and drop related PRs and after that I will probably release a 0.10.1.dev version (keeping the dev as this is not official...)

**Pull request merged**